pilot/pkg/serviceregistry/serviceentry/servicediscovery.go

```diff
@@ -503,7 +505,16 @@ func (s *ServiceEntryStore) Services() ([]*model.Service, error) {
 	s.mutex.RLock()
 	allServices := s.services.getAllServices()
 	s.mutex.RUnlock()
-	return autoAllocateIPs(allServices), nil
+
+	out := make([]*model.Service, 0, len(allServices))
+	for _, svc := range allServices {
+		// TODO: eliminate the deepcopy here
+		// autoAllocateIPs will re-allocate ips for the service,
+		// if return the pointer directly, there will be a race with `BuildNameTable`
```

Comment:
hmm.. this makes sense but is odd it didn't happen it current code?

Reply:
previously, the services are all recreated with maybeRefreshIndex each time this function is called.